packages/gatsby/src/commands/serve.js

```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
   return JSON.parse(rawJSON)
 }
 
+/**
+ * Adds trailing slash to request.url, in order to deceive static middleware
+ * into getting a directory request, thus preventing redirect.
+ * This is an express middleware to be used before static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  // url is home page
+  if (req.url === `/`) return next()
+  // url has trailing slash, no need to intervene
+  if (req.url.slice(-1) === `/`) return next()
+  // url is a file request, no need for trailing slash
+  if (fileExtensionRegEx.test(req.url)) return next()
```

Comment:
This seems to overshoot a little bit. This is basically the `.includes()` check while trying to assert that there is a non-dot before and after the dot. I don't have enough context but is there a particular reason for this?

Also, I think the `$` in the regex serves no purpose? (I don't think there's a difference with and without it). If you're set on keeping it, I think `/.\../` is the same regex. Additionally, this assumes folder parts can not have a dot (that may be legit, depending on the source of urls, but is certainly not generic).